lib/common/acl.c

```diff
@@ -1081,6 +1084,89 @@ parse_params(xsltTransformContext *transform_ctxt, const char **fallback,
     return ret;
 }
 
+static const char *
+prioritize_config_files(const char* preferred)
+{
+    int f;
+    const char *path = preferred;
+
+    /* extra override */
+    if (path != NULL && (f = open(path, O_RDONLY) != -1)) {
+        close(f);
+        return path;
+    }
+
+    /* does the user have a preference? */
+    path = pcmk__user_config(ACL_RENDER_CFG ".xsl");
+    if (path != NULL && (f = open(path, O_RDONLY)) != -1) {
+        close(f);
+        crm_debug("Sticking with user's configuration: %s", path);
+        return path;
+    }
+    free((char *) path);
+    crm_debug("Attempt to follow user's configuration failed");
+
+    /* in development mode? */
+    if (getenv("PCMK_schema_directory") != NULL) {
+        path = pcmk__xml_artefact_path(pcmk__xml_artefact_ns_base_xslt,
+                                       ACL_RENDER_CFG);
+        if (path != NULL && (f = open(path, O_RDONLY)) != -1) {
+            close(f);
+            crm_debug("Sticking with developmental configuration: %s", path);
+            return path;
+        }
+        free((char *) path);
+        crm_debug("Attempt to follow developmental configuration failed");
+    }
+
+    /* we'll go the standard path then */
+    return NULL;
+}
+
+static const char *
+gen_tmp_xml_catalog(const char *appointed_cfg, bool free_source)
+{
+    int fd;
+    FILE * f;
+    mode_t prev_umask;
+    char *catalog_path = crm_strdup_printf("%s/%s", P_tmpdir,
+                                           "pcmk-xmlcatalog-XXXXXX");
+    char *dname;
+
+    CRM_ASSERT(catalog_path != NULL);
+    CRM_ASSERT(appointed_cfg != NULL);
+
+    dname = g_path_get_dirname(appointed_cfg);
+    CRM_ASSERT(dname != NULL);
+
+    prev_umask = umask(S_IRUSR | S_IWUSR);
+    /* XXX flock(2) where possible? */
+    fd = mkstemp(catalog_path);
+    umask(prev_umask);
+
+    if ((f = fdopen(fd, "w")) == NULL) {
+        free(catalog_path);
+        return NULL;
+    }
+
+    fprintf(f, "%s%s%s\n",
+    "<catalog xmlns='urn:oasis:names:tc:entity:xmlns:xml:catalog'>\n"
+    "  <rewriteURI uriStartString='http://clusterlabs.org/nsredir/pacemaker/cfg' rewritePrefix='file://",
+            dname,
+      "'/>\n"
+      "'<nextCatalog catalog='file:///etc/xml/catalog'/>\n"
+    "</catalog>");
+
+    fclose(f);
+
+    if (free_source) {
+        free((char *) appointed_cfg);
```

Comment:
I'm not a fan of breaking const signatures; it can lead to surprises and confusion. It would be easier to follow if the caller freed the value when needed (or if it was char* and free_source was the only behavior).

Reply:
Agreed, and luckily, made me look deeper into `libxml2/catalog.c`,
leading to realization that I am going the bad route -- it doesn't
matter that much in one-off execution, but spoiling global state
of 3rd party library is wrong anyway (consider the static function
bubbling up to something widely reused, also in daemon context,
for instance).  So this whole chunk is getting an overhaul.